cloudinit/sources/DataSourceVultr.py

```diff
@@ -0,0 +1,141 @@
+
+    def __init__(self, sys_cfg, distro, paths):
+        super(DataSourceVultr, self).__init__(sys_cfg, distro, paths)
+        self.ds_cfg = util.mergemanydict([
+            util.get_cfg_by_path(sys_cfg, ["datasource", "Vultr"], {}),
+            BUILTIN_DS_CONFIG])
+        BUILTIN_DS_CONFIG['url'] = self.ds_cfg.get(
+            'url', BUILTIN_DS_CONFIG['url'])
+        BUILTIN_DS_CONFIG['retries'] = self.ds_cfg.get(
+            'retries', BUILTIN_DS_CONFIG['retries'])
+        BUILTIN_DS_CONFIG['timeout'] = self.ds_cfg.get(
+            'timeout', BUILTIN_DS_CONFIG['timeout'])
+        BUILTIN_DS_CONFIG['wait'] = self.ds_cfg.get(
+            'wait', BUILTIN_DS_CONFIG['wait'])
+
+    # Initiate data and check if Vultr
+    def _get_data(self):
+        LOG.debug("Detecting if machine is a Vultr instance")
+        if not vultr.is_vultr():
+            LOG.debug("Machine is not a Vultr instance")
+            return False
+
+        LOG.debug("Machine is a Vultr instance")
+
+        config = vultr.generate_config(BUILTIN_DS_CONFIG)
+
+        # Dump vendor config so diagnosing failures is manageable
+        LOG.debug("Vultr Vendor Config:")
+        LOG.debug(json.dumps(config))
+
+        md = self.get_metadata()
+
+        self.metadata_full = md
+        self.metadata['instanceid'] = self.metadata_full['instanceid']
+        self.metadata['local-hostname'] = self.metadata_full['hostname']
+
+        # Default hostname is "vultr"
+        if self.metadata['local-hostname'] == "":
+            self.metadata['local-hostname'] = "vultr"
+
+        self.metadata['public-keys'] = md["public-keys"].splitlines()
+        self.userdata_raw = md["user-data"]
+        if self.userdata_raw == "":
+            self.userdata_raw = None
+        self.vendordata_raw = "#cloud-config\n%s" % json.dumps(config)
+
+        # Dump some data so diagnosing failures is manageable
+        LOG.debug("SUBID: %s", self.metadata['instanceid'])
+        LOG.debug("Hostname: %s", self.metadata['local-hostname'])
+        if self.userdata_raw is not None:
+            LOG.debug("User-Data:")
+            LOG.debug(self.userdata_raw)
+
+        return True
+
+    # Get the metadata by flag
+    def get_metadata(self):
+        return vultr.get_cached_metadata(BUILTIN_DS_CONFIG)
+
+    # Compare subid as instance id
+    def check_instance_id(self, sys_cfg):
+        if not vultr.is_vultr():
+            return None
+
+        # Baremetal has no way to implement this in local
+        if vultr.is_baremetal():
+            return None
+
+        subid = vultr.get_sysinfo()['subid']
+        return sources.instance_id_matches_system_uuid(subid)
+
+    # Currently unsupported
+    @property
+    def launch_index(self):
+        return None
+
+    # Write the base configs every time. These are subject to change
+    @property
+    def network_config(self):
+        config = vultr.generate_network_config(BUILTIN_DS_CONFIG)
+        config_raw = json.dumps(config)
```

Comment:
this seems unnecessarily loud to log the dumpped config here. but if you wanted to do it, just skip the 'config_raw'  assignment and use 'json.dumps()` and do something like:

LOG.debug("Generated Network: %s", json.dumps(config))